File: 146.LRU_Cache.java
class LRUCache {

    int capacity;
    Node head;
    Node tail;
    Map<Integer, Node> hm;
    
    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.hm = new HashMap<>();
        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }
    
    public int get(int key) {
        if (hm.containsKey(key)) {
            Node curr = hm.get(key);
            remove(curr);
            insert(curr);
            return curr.value;
        }
        return -1;
    }
    
    public void put(int key, int value) {
        if (hm.containsKey(key)) {
            remove(hm.get(key));
        }
        Node newNode = new Node(key, value);
        hm.put(key, newNode);
        insert(newNode);
        if (hm.size() > this.capacity) {
            Node lastNode = this.head.next;
            remove(lastNode);
            hm.remove(lastNode.key);
        }
    }
    
    public void remove(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
    
    public void insert(Node node) {
        Node lastNode = this.tail.prev;
        lastNode.next = node;
        node.prev = lastNode;
        node.next = this.tail;
        this.tail.prev = node;
    }
    
}

class Node {
    int key;
    int value;
    Node prev;
    Node next;

    protected Node(int key, int value) {
        this.key = key;
        this.value = value;
    }
}



/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146.LRUCache.py
class Node:
    def __init__(self, key, val):
        self.val = val
        self.next = None
        self.prev = None
        self.key = key

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.dict = dict()
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head
        
    def get(self, key: int) -> int:
        if key in self.dict:
            node = self.dict[key]
            self.remove(node)
            self.add(node)
            return node.val
        return -1
        

    def put(self, key: int, value: int) -> None:
        if key in self.dict:
            self.remove(self.dict[key])
        node = Node(key, value)
        self.dict[key] = node
        self.add(node)
        if len(self.dict) > self.capacity:
            n = self.head.next
            self.remove(n)
            del self.dict[n.key]

            
    def remove(self, node):
        p = node.prev
        n = node.next
        p.next = n
        n.prev = p
    
    def add(self, node):
        p = self.tail.prev
        p.next = node
        node.prev = p
        node.next = self.tail
        self.tail.prev = node
    


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)